openeo/extra/job_management/__init__.py

```diff
@@ -606,27 +606,43 @@ def _launch_job(self, start_job, df, i, backend_name, stats: Optional[dict] = No
                 df.loc[i, "status"] = "skipped"
                 stats["start_job skipped"] += 1
 
+
     def on_job_done(self, job: BatchJob, row):
         """
         Handles jobs that have finished. Can be overridden to provide custom behaviour.
 
         Default implementation downloads the results into a folder containing the title.
+        Default implementation runs the download in a separate thread.
 
         :param job: The job that has finished.
         :param row: DataFrame row containing the job's metadata.
         """
-        # TODO: param `row` is never accessed in this method. Remove it? Is this intended for future use?
+        _log.info(f"Job {job.job_id} completed. Preparing to handle completion.")
 
         job_metadata = job.describe()
         job_dir = self.get_job_dir(job.job_id)
         metadata_path = self.get_job_metadata_path(job.job_id)
-
         self.ensure_job_dir_exists(job.job_id)
-        job.get_results().download_files(target=job_dir)
 
+        # Save metadata
+        _log.info(f"Saving metadata for job {job.job_id} to {metadata_path}")
         with metadata_path.open("w", encoding="utf-8") as f:
             json.dump(job_metadata, f, ensure_ascii=False)
 
+        # Define download logic inline
+        def download_task():
+            try:
+                _log.info(f"Starting download for job {job.job_id} to directory {job_dir}")
+                job.get_results().download_files(target=job_dir)
```

Comment:
you use `job` here as non-local variable from the `on_job_done` closure. I wonder how thread-safe this is.

A batch job object on it's own is not that complex except for the `connection` inside it: that's quite a complex object with a lot of state and risk of being thread-unsafe.  For robust threaded downloading I would  try to minimize the data structures you are sharing from the main thread to the download thread. E.g. just the backend URL (a string), the job id (a string) and a valid access token (a string) might be enough. Or even better: just pass the download URLs if they are signed (and don't require a working access token)